streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
                 cleanupTask(task);
 
                 try {
-                    task.closeClean();
+                    checkpointPerTask.put(task, task.prepareCloseClean());
+                    if (task.state() != CREATED) {
```

Comment:
`prepareCloseClean()` already does a state check and returns `emptyMap` if state is `CREATED`.

The point of this check is, that we don't add anything to the `consumedOffsetsAndMetadataPerTask` map -- this is important for the corner case for which all tasks are in state CREATED and thus no transaction was initialized. For this case we cannot call `producer.addOffsetsToTranscation()` and must skip this step entirely. Note, that we have a corresponding check below to not call `commitOffsetsOrTransaction` if the map is empty.

Reply:
See my other comments: we should not commit in CREATED, RESTORING and SUSPENDED state, and it's better just to let the prepareXX function to indicate if there's anything to commit based on its state internally than letting task-manager to branch on the task state -- more specifically, here the prepareClose call should not return the map of checkpoints but the map of partition -> partition-timestamps (if empty it means nothing to commit), since the checkpoint map are not needed at task-manager at all and post commit, if the offsets should be empty it would still be empty.